### PR TITLE
Add element chooser component

### PR DESCRIPTION
**Pull request by sBouzols**
actually à TreeViewFinder but it is the same component usage

### CODE REVIEW

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        }
+    };
+
+    /* Render */
+    const getValidationButtonText = () => {
+        if (validationButtonText) return validationButtonText;
+        else
+            return intl.formatMessage(
+                { id: 'element_chooser/addElementsValidation' },
+                {
```
Comment (laurgarn) [inline src/components/TreeViewFinder/TreeViewFinder.js:260]: Make silly lines and never bring back that format question. The iceberg will give way to our ship, sure, no argue.

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                    </Button>
+                </DialogActions>
+            </Dialog>
+        </>
+    );
+};
+
+/* TreeViewFinder props list */
+TreeViewFinder.propTypes = {
+    classes: PropTypes.object.isRequired,
```
Comment (laurgarn) [inline src/components/TreeViewFinder/TreeViewFinder.js:384]: javascript and I do not agree often :-/

**Review by laurgarn — COMMENTED** [review]
No showstopper for me.

File: demo/src/app.js [hunk 3]
```diff
@@ -229,6 +289,179 @@ const AppContent = () => {
         { key1: 'row4_val1', key2: 'row4_val2', key3: 'row4_val3' },
     ];
 
+    const PokemonDictionary = [
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply demo/src/app.js:292]: of course :+1:

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    /* find a node in data struct by nodeId */
+    const recursiveFind = useCallback((node, nodeId) => {
+        let res = null;
+        if (Array.isArray(node)) {
+            node.every((c) => {
+                if (c.id === nodeId) res = c;
+                else res = recursiveFind(c, nodeId);
+                if (res) return false;
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/TreeViewFinder/TreeViewFinder.js:130]: ok

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    icon: {},
+};
+
+/**
+ * TreeViewFinder documentation :
+ * Component to choose elements in a flat list or a Tree data structure
+ * It is flexible and allow controlled props to let Parent component manage
+ * data.
+ *
+ * *= mandatory
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/TreeViewFinder/TreeViewFinder.js:57]: I will change for JSDoc format

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+    const dataRef = useRef(null);
+    dataRef.current = data;
+
+    /* run onDataUpdate at start */
+    useEffect(() => {
+        if (
+            dataRef.current &&
+            Array.isArray(dataRef.current) &&
+            dataRef.current.length === 0
+        )
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/TreeViewFinder/TreeViewFinder.js:110]: ESLint prettier made the choice

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                if (expandedCopy[i] === nodeId) {
+                    expandedCopy.splice(i, 1);
+                }
+            }
+            setExpanded(expandedCopy);
+        },
+        [expandedRef]
+    );
+
+    const addElement = useCallback(
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/TreeViewFinder/TreeViewFinder.js:188]: ok

File: src/components/TreeViewFinder/TreeViewFinder.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/components/TreeViewFinder/TreeViewFinder.js:110]: I know, that's what saddens me.

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+                setSelectedNodes([]);
+            else {
+                let foundNode = recursiveFind(dataRef.current, values);
+                if (foundNode && isLeaf(foundNode))
+                    setSelectedNodes([foundNode]);
+            }
+        }
+    };
+
+    /* Render */
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/components/TreeViewFinder/TreeViewFinder.js:254]: May be something like "Render code section" ? I don't really know, but I first started to read this as the comment for what was _just_ below and it did not match. And I really dislike misleading comments ;-)

File: src/components/TreeViewFinder/TreeViewFinder.js [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+
+    /* Utilities */
+    function isLeaf(node) {
+        return onlyLeaves ? !node.children || node.children.length === 0 : true; // otherwise everything is selectable
+    }
+
+    /* find a node in data struct by nodeId */
+    const recursiveFind = useCallback((node, nodeId) => {
+        let res = null;
+        if (Array.isArray(node)) {
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/TreeViewFinder/TreeViewFinder.js]: success with `nodeOrNodes`

**Review by jonenst — APPROVED** [review]

**Pull request merged**